src/docfx/build/toc/parser/TableOfContentsParser.cs

```diff
@@ -116,30 +129,58 @@ private static (List<Error>, TableOfContentsModel) LoadTocModel(JToken tocToken)
             Document rootPath,
             ResolveContent resolveContent,
             ResolveHref resolveHref,
-            ResolveXref resolveXref)
+            ResolveXref resolveXref,
+            ResolveMoniker resolveMoniker,
+            List<string> rootFileMonikers)
         {
             var errors = new List<Error>();
             foreach (var tocModelItem in tocModelItems)
             {
                 if (tocModelItem.Items != null && tocModelItem.Items.Any())
                 {
-                    errors.AddRange(ResolveTocModelItems(context, tocModelItem.Items, parents, filePath, rootPath, resolveContent, resolveHref, resolveXref));
+                    errors.AddRange(ResolveTocModelItems(context, tocModelItem.Items, parents, filePath, rootPath, resolveContent, resolveHref, resolveXref, resolveMoniker, rootFileMonikers));
                 }
 
                 // process
                 var tocHref = GetTocHref(tocModelItem);
                 var topicHref = GetTopicHref(tocModelItem);
                 var topicUid = tocModelItem.Uid;
 
+                var monikers = new List<string>();
+
                 var (resolvedTocHref, resolvedTopicItemFromTocHref, subChildren) = ProcessTocHref(tocHref);
                 var (resolvedTopicHref, resolvedTopicName, document) = ProcessTopicItem(topicUid, topicHref);
 
                 // set resolved href back
                 tocModelItem.Href = resolvedTocHref ?? resolvedTopicHref ?? resolvedTopicItemFromTocHref?.Href;
+                if (resolvedTocHref != null || resolvedTopicHref != null)
+                {
+                    var linkType = UrlUtility.GetLinkType(resolvedTopicHref);
+                    if (linkType == LinkType.External || linkType == LinkType.AbsolutePath)
+                    {
+                        monikers = rootFileMonikers;
+                    }
+                    else
+                    {
+                        monikers = resolveMoniker(document);
+                    }
+                }
+                else
+                {
+                    tocModelItem.Href = resolvedTopicItemFromTocHref?.Href;
+                    monikers = resolvedTopicItemFromTocHref?.Monikers;
+                }
+
                 tocModelItem.TocHref = resolvedTocHref;
                 tocModelItem.Name = tocModelItem.Name ?? resolvedTopicName;
                 tocModelItem.Items = subChildren?.Items ?? tocModelItem.Items;
 
+                // Union with children's monikers
+                var childrenMonikers = tocModelItem.Items?.SelectMany(child => child?.Monikers ?? new List<string>()) ?? new List<string>();
```

Comment:
did we initialize items and monkers to empty list? these ? and ?? null checks are clumsy

Reply:
with `TocModelItem { public string[] Monikers { get; set; } = Array.Empty<string>() }`, this could be:

```suggestion
                var childrenMonikers = tocModelItem.Items.SelectMany(child => child.Monikers);
```